### PR TITLE
Feature/153 bx can peripheral abstraction

### PR DESCRIPTION
**Pull request by crsz20**
### Summary
Abstracts away the BxCAN (Basic Extended) peripheral from ST's HAL library.
 The intention is that the ECU firmware component will interact through an abstract interface, `ICan`. That is, the ECU component should not care about which platform it is running on. Changing to another platform or hardware library will require the developer to implement their concrete class of the `ICan` interface.

### Remarks
- The `app.cpp` file shows how to integrate the `BxCanStmF4` class.
- This is our first use case for using `std::shared_pointer`, instead of simply using `std::unique_pointer`. This is because the interrupt callback function `HAL_CAN_RxFifo0MsgPendingCallback()` needs to be in global scope to override the HAL library's function. In other words, both the global scope and the `BxCanStmF4` need access to the same `hcan1` peripheral.
- Overriding is ST's expectation of defining the interrupt callback function, as the source code is declares it as a "weak" function — not defines it.
  - In other words: `__weak void HAL_CAN_RxFifo0MsgPendingCallback(CAN_HandleTypeDef *hcan)`
  - This is found on line 2,236 of ST's `stm32f4xx_hal_can.c` file.

### Checks
- [x] Tested

### CODE REVIEW

File: Project/DAQ_System/Program/Src/Platform/CAN/Interfaces/ican.hpp [hunk 1]
```diff
@@ -12,12 +12,30 @@
...
+
+	virtual bool MessageArrivedFlag() = 0;
+
+	virtual void ClearMessageArrivedFlag() = 0;
+
+	virtual uint32_t LatestCanId() = 0;
+
+	virtual void EnableInterruptMode() = 0;
 
+	virtual void DisableInterruptMode() = 0;
```
Comment (WilliamLim781) [inline Project/DAQ_System/Program/Src/Platform/CAN/Interfaces/ican.hpp]: This isn't relevant to what we are doing but having another function in the abstract interface where you can set custom ID masks for the CAN Controller filtering might be a nice addition to the abstract interface.

File: Project/DAQ_System/Program/Src/Platform/CAN/STM/F4/bxcan_stmf4.cpp [hunk 2]
```diff
@@ -15,7 +15,93 @@
...
+	for (uint8_t i = 0; i < can_id_list.size(); i++) {
+		ConfigureFilter(filter, i, can_id_list.at(i));
+	}
 
+	HAL_CAN_Start(&bx_can_);
+}
+
+void BxCanStmF4::ConfigureFilter(CAN_FilterTypeDef &filter, uint8_t filter_bank_num, uint32_t can_id) {
+	filter.FilterIdHigh = can_id >> 13 ;
+	filter.FilterIdLow = ((can_id & 0x1FFF) <<3) | CAN_ID_EXT | CAN_RTR_DATA ;
```
Comment (WilliamLim781) [inline Project/DAQ_System/Program/Src/Platform/CAN/STM/F4/bxcan_stmf4.cpp:35]: Just making a note that this probably should be documented later since it CAN(lol) be rather confusing what is going on here

**Review by WilliamLim781 — APPROVED** [review]
Honestly great work on this! It's great to see progress on the DAQ especially with something as big as CAN. :thumbsup:

File: Project/DAQ_System/Program/Src/Platform/CAN/Interfaces/ican.hpp (hunk 1 above)
Comment (crsz20) [reply Project/DAQ_System/Program/Src/Platform/CAN/Interfaces/ican.hpp]: Done. The user of the library is now expected to do the following:
- Provide the BxCAN peripheral from the library through the constructor
- Provide a list of CAN IDs
- Start the peripheral
- Configure the interrupt for the receive callback
- Enable the interrupt
- Profit

**Comment by crsz20** [comment]
I will retest my latest changes on the car when I test the ECU class (should be my next PR after this one).

File: Project/DAQ_System/Program/Src/Platform/CAN/Interfaces/ican.hpp [hunk 3]
```diff
@@ -12,12 +12,35 @@
...
+// Standard Libraries
+#include <cstdint>
+#include <vector>
+
 namespace platform {
 
 class ICan {
-    public:
-        virtual ~ICan() = default;
+public:
```
Comment (alexshidagoatnocap) [inline Project/DAQ_System/Program/Src/Platform/CAN/Interfaces/ican.hpp:22]: Indent this line

File: Project/DAQ_System/Program/Src/Platform/CAN/STM/F4/bxcan_stmf4.hpp [hunk 4]
```diff
@@ -14,10 +14,70 @@
...
+#include "can.h"
+
+// DFR Custom Dependencies
+#include "../../Interfaces/ican.hpp"
+
+
 namespace platform {
 
-class BxCanStmF4 {
+class BxCanStmF4 : public ICan {
```
Comment (alexshidagoatnocap) [inline Project/DAQ_System/Program/Src/Platform/CAN/STM/F4/bxcan_stmf4.hpp:29]: It may just be me but what does this line do? Why does it have a ` : public ICan` after the class definition?

**Review by alexshidagoatnocap — COMMENTED** [review]
Code is well written and easy to read

File: Project/DAQ_System/Program/Src/Platform/CAN/STM/F4/bxcan_stmf4.hpp (hunk 4 above)
Comment (WilliamEverson) [reply Project/DAQ_System/Program/Src/Platform/CAN/STM/F4/bxcan_stmf4.hpp:29]: This makes the class BxCanStmF4 inherit the member functions of ICan

File: Project/DAQ_System/Program/Src/app.cpp
Comment (WilliamEverson) [inline Project/DAQ_System/Program/Src/app.cpp:1]: There seem to be a lot of code commented out in this file. I would assume some of it is unit testing (Lines 111 - 132), are any these lines necessary?

Comment (crsz20) [reply Project/DAQ_System/Program/Src/app.cpp:1]: Those commented lines are of our other firmware components: linear potentiometer, accelerometer, and gyroscope. I commented them for the sake of keeping the integration.

Unit tests are held in the [Tests Directory](https://github.com/DallasFormulaRacing/DataAcquisition2.0/tree/develop/Project/Tests). If you're interested in how it works, checkout the [Unit Testing](https://github.com/DallasFormulaRacing/DataAcquisition2.0/wiki/Unit-Testing) page.

File: Project/DAQ_System/Program/Src/Platform/CAN/Interfaces/ican.hpp (hunk 3 above)
Comment (crsz20) [reply Project/DAQ_System/Program/Src/Platform/CAN/Interfaces/ican.hpp:22]: I am indifferent when it comes to indenting `public:` and `private` lines. CubeIDE seems to revert it to that way, so I leave it alone.

**Review by alexshidagoatnocap — APPROVED** [review]

**Review by WilliamEverson — APPROVED** [review]
